src/libraries/System.Private.CoreLib/src/System/Enum.EnumInfo.cs

```diff
@@ -5,19 +5,47 @@ namespace System
 {
     public abstract partial class Enum
     {
-        internal sealed class EnumInfo
+        internal sealed unsafe class EnumInfo
         {
             public readonly bool HasFlagsAttribute;
             public readonly ulong[] Values;
             public readonly string[] Names;
 
+            // The cached managed function pointer to the right JIT allocation helper.
+            // Same as in RuntimeType.ActivationCache, just copied here to avoid cache conflicts.
+            private delegate*<void*, object> _pfnAllocator;
+            private void* _allocatorFirstArg;
+
             // Each entry contains a list of sorted pair of enum field names and values, sorted by values
             public EnumInfo(bool hasFlagsAttribute, ulong[] values, string[] names)
             {
                 HasFlagsAttribute = hasFlagsAttribute;
                 Values = values;
                 Names = names;
             }
+
+            /// <summary>
+            /// Creates an uninitialized object of the current type.
+            /// </summary>
+            /// <returns>An uninitalized instance of the object of the specified type.</returns>
+            internal object CreateUninitializedInstance(RuntimeType type)
+            {
+                Diagnostics.Debug.Assert(type.IsEnum);
+
+                if (_pfnAllocator is null)
```

Comment:
This does not look thread-safe.

What guarantees that `_allocatorFirstArg` is initialized?

Reply:
It's not thread-safe, I just assumed in the worst case scenario the FCall would be repeated, but no harm done? From what I can see, other caching mechanisms for `RuntimeType` also do the same, with the cache potentially being initialized multiple times in case there's a race condition. I assume that that QCall is just returning the fixed address of the right JIT allocation helper, so there's no allocations involved and no concerns in case the call is made multiple times? 🤔

Reply:
This is initializing two values: `_pfnAllocator` and `_allocatorFirstArg`. You want to make sure that both of these values are initialized before they are used below. This condition does not guarantee that both values are initialized.

Consider this interleaving:
Thread 1: Initializes `_pfnAllocator`
Thread 2: `if (_pfnAllocator is null)` condition is false, so the initialization is skipped
Thread 2: Executes `_pfnAllocator(_allocatorFirstArg);`. The allocator is going to crash since `_allocatorFirstArg` is not yet initalized.
Thread 1: Initializes `_allocatorFirstArg`.